### ISSUE TITLE
ansible 2.8.6 get_url repeatedly downloads file with force==no

### ISSUE THREAD

**Issue by gtie** [ansible/ansible#64016]
##### SUMMARY
Unlike in version 2.8.5, get_url module repeatedly downloads an URL even if destination exists with Ansible 2.8.6

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
get_url

##### ANSIBLE VERSION

```
$ ansible --version
ansible 2.8.6
  config file = None
  configured module search path = [u'/Users/XXX/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/XXX/.virtualenvs/ansible2.8/lib/python2.7/site-packages/ansible
  executable location = /Users/XXX/.virtualenvs/ansible2.8/bin/ansible
  python version = 2.7.16 (default, Sep  2 2019, 11:59:44) [GCC 4.2.1 Compatible Apple LLVM 10.0.1 (clang-1001.0.46.4)]
```

##### CONFIGURATION

```
$ ansible-config dump --only-changed
$
```

##### OS / ENVIRONMENT
MacOS

##### STEPS TO REPRODUCE

```yaml
$ touch foofile
$ cat > test_playbook.yaml
---
- hosts: localhost
  connection: local

  tasks:
    - name:           Fetch file
      get_url:
        url:          http://google.com
        dest:         ./foofile
        force:        no

$ ansible-playbook -i 'localhost,' test_playbook.yaml
```

##### EXPECTED RESULTS
No changes, since "foofile" already exists. With the same playbook, a run with Ansible 2.8.5 results in:
```
PLAY RECAP ********************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

##### ACTUAL RESULTS
File gets overwritten/downloaded every time
```
$ ansible-playbook -i 'localhost,' test_playbook.yaml
PLAY [localhost] **************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************
ok: [localhost]

TASK [Fetch file] *************************************************************************************************
changed: [localhost]

PLAY RECAP ********************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/net_tools/basics/get_url.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/net_tools/basics/get_url.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @jpmens @ptux
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by sivel** [comment]
There was a bug in <2.8.6 where we assumed no checksum was a checksum match.  I think we need to further improve the logic to account for this situation too.

Maybe something like:

```diff
diff --git a/lib/ansible/modules/net_tools/basics/get_url.py b/lib/ansible/modules/net_tools/basics/get_url.py
index 3c84cb60e9..3fb72dbe84 100644
--- a/lib/ansible/modules/net_tools/basics/get_url.py
+++ b/lib/ansible/modules/net_tools/basics/get_url.py
@@ -540,7 +540,7 @@ def main():
                 checksum_mismatch = True
 
         # Not forcing redownload, unless checksum does not match
-        if not force and checksum and not checksum_mismatch:
+        if not force and (not checksum or (checksum and not checksum_mismatch)):
             # Not forcing redownload, unless checksum does not match
             # allow file attribute changes
             module.params['path'] = dest
```

More integration tests would need added.

**Comment by telbizov** [comment]
@sivel I tested the proposed patch above and it seems like the logic is fixed and matches the intended behaviour. If the file exists locally, regardless of its contents if skips the download.

What version of ansible do you envision to incorporate the fix in?

**Comment by gtie** [comment]
@sivel, the condition fix looks good - how soon can it get released? With a large server farm, this bug triggers a stampede effect, where there would be 0 requests normally.

**Comment by sivel** [comment]
I am not actively working on this issue.  I have provided a proof of concept fix that someone else can validate, make modifications if necessary, provide a pull request to resolve it, and include integration tests.

**Comment by hugocartwright** [comment]
@gtie @sivel I am currently looking into an implementation of @sivel 's POC and am adding integration tests for this issue too. Currently checking all previous integration tests are not impacted by the change I want to propose.

**Comment by gtie** [comment]
@hugocartwright, what needs to happen for your PR to get merged? Also, can it get backported to the 2.8 branch?

**Comment by hugocartwright** [comment]
@telbizov @gtie Integration tests fail for me with @sivel 's suggestion. An extra flag required besides "not checksum"? Regardless, I am PRing it with some suggestions for progress.

The test failures seem to come from the following commit:
- https://github.com/ansible/ansible/commit/7d51cac3301fddf6a876b2f75057f432b0841f04#diff-db467d0b6402f54d9beb7ec4ae4b0c5d

@mscherer @pilou- Thanks

**Comment by gtie** [comment]
@hugocartwright, is perhaps the failing test a consequence of changing the test URL from `https://{{ httpbin_host }}/get` to `https://{{ httpbin_host }}/html`? What is the purpose of the URL change so far away from the actual double-download assertion (mid-file vs end-of-file)?

**Comment by whs-dot-hk** [comment]
Hi, can I start working on this issue?

**Comment by Derioss** [comment]
Same issue with docker image debian:testing-slim (python2.7) and python:3.7.6-slim-buster.
Works on ansible 2.8.4 and 2.8.5
don't works on 2.8.6 +

This change seems to be the issue
We never get into the conditional block because of the `checksum`
```python
# Not forcing redownload, unless checksum does not match
if not force and checksum and not checksum_mismatch:
    # Not forcing redownload, unless checksum does not match
    # allow file attribute changes
    file_args = module.load_file_common_arguments(module.params, path=dest)
    result['changed'] = module.set_fs_attributes_if_different(file_args, False)
    if result['changed']:
        module.exit_json(msg="file already exists but file attributes changed", **result)
    module.exit_json(msg="file already exists", **result)
```

**Comment by dmrzzz** [comment]
@sivel I don't see why checksum enters into the desired logic at all.  Even if I did provide a checksum to guard against bad downloads, force:no means I _never_ want to overwrite my existing file.

Documentation of `force` in https://docs.ansible.com/ansible/latest/modules/get_url_module.html#parameter-force says "If no, the file will only be downloaded if the destination does not exist." which is clear, intuitive, and consistent with the meaning of `force` in other contexts -- does the copy module with force:no overwrite an existing file which doesn't match the checksum?  (okay, actually I've never tried that, but I really hope not)

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/get_url.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/get_url.py)

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by vladimir-mencl-eresearch** [comment]
Hi,
I've been watching this issue (and the related PRs like #65340) for a while and I think there is confusion (at least between code and documentation) over what should happen if:
* no `checksum` is given
* `force` is set to `no`
* `dest` exists, but differs from the content at `url`

The documentation for `force` clear says:
> If no, the file will only be downloaded if the destination does not exist.

However, the changes introduced in 2.8.6 (in 7d51cac3301fddf6a876b2f75057f432b0841f04 / #62146) ) change the behaviour to effectively always force a new download - even the commit message for this commit says:
>  Don't treat no checksum as a checksum match

I think we need a clear answer on what should happen in this case (`force: no`, no checksum, file exists but differs) - either:
(a) force a new download and replace the file with contents at URL
(b) keep the file as is

Once we have that answer, fixing the logic should be easy.

Can someone in the right position (module owner?) please pick the desired behaviour?

Cheers,
Vlad

**Comment by computerlyrik** [comment]
Please fix this - in code or documentation.

(b) is more what i expect reading the documentation
(a) is more an "ansible way" of setting a state ... imho

My workflow as follows:
- download a text file (get_url)
- replace a line (lineinfile)

So this will lead to change twice every time i execute the tasks.
With (b) it would work out of the box
With (a) i need to put in an additional file exists check (what i will do now)

**Comment by tommarcoen** [comment]
Is anyone making any progress on this? I would love to see this fixed as well, or at least have a work-around.

**Comment by vladimir-mencl-eresearch** [comment]
My (sad) workaround is to do a stat first, and skip the task if the file already exists...

**Comment by gloomytrousers** [comment]
You can work around it with a `when:` like this:
```yaml
- name: Example
  get_url:
    url: https://example.com/file
    dest: /path/to/dest
  when: not '/path/to/dest' is exists # Workaround
```
A more complex example:
```yaml
- name: Example with vars
  get_url:
    url: "{{ item.url }}"
    dest: /path/to/{{ item.dest }}
  when: not ('/path/to/' + item.dest) is exists # Workaround
  with_items:
    - { url: "https://example.com/file1", dest: file1 }
    - { url: "https://example.com/file2", dest: file2 }

```

**Comment by vladimir-mencl-eresearch** [comment]
@gloomytrousers , thanks for the suggestion, but that won't unfortunately work: the `exists` test applies to local files on the Ansible *controller* machine, not the target host.  To check for presence on the target host, an explicit `stat` task is required...

**Comment by gloomytrousers** [comment]
@vladimir-mencl-eresearch thanks for the info - we learn something new every day :-) Yes, in my case I'm running the playbook against the local machine so this works for me.

**Comment by DanaLacoste** [comment]
I am seeing this too, and it is difficult because I have some code that runs (depending on the situation) sometimes with 'net access, and sometimes without.

So when the web proxy is down, the code fails because it is trying to download something that is already there (and we are upgrading from pre-2.8.5 to 2.9.x and running into this)

I am going to wrap a complicated check around to make the `when:` reliable, but it would be nice to make this work!

Thanks!

**Comment by ptux** [comment]
@sivel 

You could assign this to me.
I'm going to fix it.

**Comment by tommarcoen** [comment]
I am currently using @vladimir-mencl-eresearch 's workaround using `stat` but am still curious about a permanent fix. Is anyone working on it? If not, why not? Can I help?

**Comment by Daryes** [comment]
Bumping this one as this still happens in ansible latest (v2.13.3 tag) and v2.9.27.  

Given the current change of apt-key on Debian and related dists, I implemented the use of get_url instead of apt-key to retrieve the repositories keys.  
For more information: #78063 

To my surprise, there was a http return code in the results, and yes, the request was really sent  

Test script:
```
- set_fact:
    arg_install_ppa:
      - name: "grafana-oss"
        apt_filename: "grafana.list",
        apt_key_url: "https://packages.grafana.com/gpg.key",
        apt_repository: "deb https://packages.grafana.com/oss/deb stable main",

- name: "{{ arg_install_comment }} | pkg install APT - add ppa keys via gpg - key installation"
  get_url:
    url: "{{ item.apt_key_url }}"
    dest: "/etc/apt/trusted.gpg.d/{{ item.name }}.repo.asc"
    backup: no
    force: no
  with_items:
    - "{{ arg_install_ppa }}"
  register: task_update_key_gpg

- debug: var=task_update_key_gpg
```

Result: 
```
TASK [pkg_install : debug] ********************************************************************************************************************************************************************
ok: [prom] => {
    "task_update_key_gpg": {
        "changed": false,
        "msg": "All items completed",
        "results": [
            {
                "ansible_loop_var": "item",
                "changed": false,
                "dest": "/etc/apt/trusted.gpg.d/grafana-oss.repo.asc",
                "elapsed": 0,
                "failed": false,
                "gid": 0,
                "group": "root",
                "invocation": {
                    "module_args": {
                        "attributes": null,
                        "backup": false,
                        "checksum": "",
                        "client_cert": null,
                        "client_key": null,
                        "content": null,
                        "delimiter": null,
                        "dest": "/etc/apt/trusted.gpg.d/grafana-oss.repo.asc",
                        "directory_mode": null,
                        "follow": false,
                        "force": false,
                        "force_basic_auth": false,
                        "group": null,
                        "headers": null,
                        "http_agent": "ansible-httpget",
                        "mode": null,
                        "owner": null,
                        "regexp": null,
                        "remote_src": null,
                        "selevel": null,
                        "serole": null,
                        "setype": null,
                        "seuser": null,
                        "sha256sum": "",
                        "src": null,
                        "timeout": 10,
                        "tmp_dest": null,
                        "unsafe_writes": false,
                        "url": "https://packages.grafana.com/gpg.key",
                        "url_password": null,
                        "url_username": null,
                        "use_proxy": true,
                        "validate_certs": true
                    }
                },
                "item": {
                    "apt_filename": "grafana.list",
                    "apt_key_url": "https://packages.grafana.com/gpg.key",
                    "apt_repository": "deb https://packages.grafana.com/oss/deb stable main",
                    "name": "grafana-oss"
                },
                "mode": "0644",
                "msg": "HTTP Error 304: Not Modified",
                "owner": "root",
                "size": 1694,
                "state": "file",
                "uid": 0,
                "url": "https://packages.grafana.com/gpg.key"
            }
        ]
    }
}
```

Notice the ` "msg": "HTTP Error 304: Not Modified",` near the end of the output. 

Given the current code [in devel](https://github.com/ansible/ansible/blob/ece85abbc46e087187caf6e05b1515b97c578531/lib/ansible/modules/get_url.py#L567)  
- the statement ` and checksum` should be removed from this line `  if not force and checksum and not checksum_mismatch:`
- final statement : `  if not force and not checksum_mismatch:`
This will be valid as long as `checksum_mismatch` stays initialized to `false` 10 lines upper

As actually get_url might become the replacement for apt-keys when a file is used as a key signature, this really needs to be fixed, as get_url will connect to the remote server each time and download the file if changed.
Which is not always welcome, it should be a conscious update than a hidden one. 

And I do agree a checksum should be used also ... just need to retrieve the key first for each external repo, and generate the hash because it is not directly provided in the products documentation. Why would it matter, the usual doc suggests to pipe the  curl of a script to sudo bash ...  :'(

**Comment by hsoft** [comment]
Let's recap, shall we?

This issue has been opened for a little while and seems to have fallen into a crack with #64626 waiting for contributor feedback and #65340 falsely considering the issue fixed (it's true that the task correctly registers as "not changed", but `fetch_url()` is still called).

It seems to me that, as @vladimir-mencl-eresearch correctly mentions, the heart of the matter is that behavior began diverging from documentation with #62146 which fixed another problem. It's possible that #62146 is the correct behavior, but in that case, it should be accompanied by a documentation update. Decision needed.

To ease the decision, I've submitted a new PR which applied cleanly to the devel branch that essentially reverts #62146 and adjusts tests accordingly.

**Comment by mackuba** [comment]
Hey, so I've just found this issue, after being confused too… @hsoft - yeah, what you wrote and your PR at #79518 makes sense to me.

But I started digging into the original PR at #62146 (pinging the author: @sivel) and the corresponding issue #61978 to see why that change has been made, and I think I understand the whole problem now. The problem is that "force" across this module kind of means two different things:

1. Whether it should force the download if file exists
2. How it should set cache headers for the download, and if it should force the download ignoring the cache (instead of allowing the download to return 304)

@sivel probably understood "force" as only 2), and expected the command to always try to perform the download if force == false, with force == true only changing that it forces the file contents to actually be downloaded instead of allowing 304. We apparently understand "force" as 1). The question is which of those should it mean (it seems to me that docs currently imply 1), and if we perhaps need another option to differentiate between these two?

If "force" means only 2, there is no way to completely skip any other checks if *some* file file at that path as we expected, other than adding additional workarounds. If "force" means only 1, then there is no way to always make a request to check for a newer version of the file allowing a 304 response.

**Comment by outerLeitmotiv** [comment]
Any update on this ? It was created 4 years ago. We are waiting since 2022(!) for a decision (!) if either the code or the doc needs to be updated.

**Comment by binaryDiv** [comment]
Please make a decision about the intended behaviour of the force parameter and/or whether to add a new parameter to support both the "update file if changed" and "only download if not existant" use cases, so that someone can create a PR for this.

I keep stumbling over this issue and until just now I thought it's an issue with check mode (see https://github.com/ansible/ansible/issues/65687), but no, the module actually doesn't work in the documented way.

**Comment by Daryes** [comment]
For the record,  this behavior is easy to circumvent with a presence test on the local file first.
But **only after** you have some understanding of the `force` parameter behavior and the incomplete documentation.
So losing much time having to debug get_url beforehand.

````
- stat:
    path: /my/dest/file
    get_checksum: no
  register: task_geturl_file
  changed_when: task_geturl_file.stat.exists == false

- get_url:
    dest: /my/dest/file
    ...
  when: task_geturl_file is changed
````

**Comment by Daryes** [comment]
You're right, as this requires a good knowledge of ansible to handle this situation.  
Having get_url working as expected/desired would be much simpler.

If you don't already have something for this, there are currently 2 working options : 

Option 1 :
The least cumbersome, but still more than a single loop on a get_url task : looping over an include_tasks. 
The generated "item" variable from the loop is usable in the included task file, so it is possible to do a stat + get_url on a single file. This will make the code much easier to read.

```
# main task file 
- include_tasks: get_url_with_presence_test.yml
  with_items:
    - "url1"
    - "url2"
    ...
  # rename the generated item to something else to prevent any conflict
  loop_control:
    loop_var: urlItem
```

```
---
# dedicated file : get_url_with_presence_test.yml

- name: "Presence check of {{ urlItem |basename }}"
  stat:
    path: "/my/dest/{{ urlItem |basename }}"
    get_checksum: no
  register: task_geturl_file
  changed_when: task_geturl_file.stat.exists == false

- name: "Retrieving the file {{ urlItem |basename }}"
  get_url:
    url: "{{ urlItem }}"
    dest: "/my/dest/{{ urlItem |basename }}"
      ...
  when: task_geturl_file is changed
```

The downside is on the log output when executed, it will add 2 tasks per url instead of a single one for the whole list.

---

Option 2 :
Not really desired, but can be useful to know this is possible for special cases :
the registered variable from the output of a loop will have its data under a `.result[ ]` array.
But all the items listed in a loop will keep the same order in the result array : the first item listed in the task will be at index 0, the second at index 1, and so on, item n will be at index n-1.
With this, the desired element can be selected without error from the registered variable output. 
```
- stat:
    path: "/my/dest/{{ item |basename }}"
    get_checksum: no
  with_items
    - "url1"
    - "url2"
    ...
  register: task_geturl_file
  # this should work even if in a the loop
  changed_when: task_geturl_file.stat.exists == false

- get_url:
    path: "/my/dest/{{ item.url |basename }}"
    ...
  with_items
    - { url: "url1", when: "{{ task_geturl_file.results.0.stat.exists == false }}" }
    - { url: "url2", when: "{{ task_geturl_file.results.1.stat.exists == false }}" }
    - { url: "url3", when: "{{ task_geturl_file.results.2.stat.exists == false }}" }
    ..
  when: item.when |bool
```
It is something I'm using in some rare cases. 
But it gives me a headache each time I need to change something in it.

**Comment by rlaager** [comment]
> Hi, I've been watching this issue (and the related PRs like [#65340](https://github.com/ansible/ansible/pull/65340)) for a while and I think there is confusion (at least between code and documentation) over what should happen if:
> 
> * no `checksum` is given
> * `force` is set to `no`
> * `dest` exists, but differs from the content at `url`

The module can't tell if it differs without downloading it (well, technically, making an HTTP request, but let's not get into HEAD requests). So, IMHO, any talk of whether it differs is irrelevant.

The point of `force: no`, as documented (and as @dmrzzz [said](https://github.com/ansible/ansible/issues/64016#issuecomment-601955021) is consistent with the `copy` module), is to _not_ download if the file already exists.

> However, the changes introduced in 2.8.6 (in [7d51cac](https://github.com/ansible/ansible/commit/7d51cac3301fddf6a876b2f75057f432b0841f04) / [#62146](https://github.com/ansible/ansible/pull/62146)) ) change the behaviour to effectively always force a new download

Those changes are wrong. They change the module to no longer match its documentation (as per this issue). If that was intended, the PR should have updated the documentation. Additionally, the code no longer matches the immediately adjacent comments. Again, if this was intentional, those comments should have been updated.

That PR was trying to fix issue #61978. In the issue, the complaint is that `get_url` is not doing a conditional GET. The error was expecting that without setting `force: yes`, i.e. keeping the default of `force: no`. With `force: yes`, it seems reasonable that it should do a conditional GET (and reading the code, I suspect it does). A conditional GET can potentially save an unnecessary download if the file has not changed server side. However, if the file exists and there is a checksum mismatch, then it should do an unconditional GET, since we know the file is wrong. Again, there is existing code that seems to do this.

I think someone (which could be me, if it looks like there is agreement on the desired behavior) should make the following changes:

- Revert the change to "Assert that we didn't re-download unnecessarily" from #62146. It should not download at all, so a 304 is not allowed.
- Fix this bug: if `force: no`, the file will not be downloaded if it exists. This will involve a revert of the change from #62146. Document this with a changelog snippet.
- Verify (and add/modify tests to check) that with `force: yes`, it is still doing a conditional GET. Except that a checksum mismatch should do an unconditional GET.
- Stretch goal: Fix the documentation, tests, and behavior to not differentiate between `dest` being a file and `dest` being a directory. That is, if `dest` is a directory, use the directory plus the basename of the URL, and check whether that file exists. Multiple people have mentioned this as being strange and inconsistent with the `uri` module.

**Comment by vladimir-mencl-eresearch** [comment]
Thanks for the initiative @rlaager !

Thumbs up from me for the proposed work.
